### ISSUE TITLE
Compiler panic, attempt to divide by zero

### ISSUE THREAD

**Issue by pontushojer** [typst/typst#3208]
### Description

I run into this error while trying to compile my file.

```
% typst compile thesis.typ 
thread 'main' panicked at /Users/brew/Library/Caches/Homebrew/cargo_cache/registry/src/index.crates.io-6f17d22bba15001f/citationberg-0.1.1/src/lib.rs:693:23:
attempt to divide by zero
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

With RUST_BACKTRACE=full I get this stack

```
stack backtrace:
   0:        0x10bd2a742 - __mh_execute_header
   1:        0x10b99aa1b - __mh_execute_header
   2:        0x10bd067fe - __mh_execute_header
   3:        0x10bd2a4f9 - __mh_execute_header
   4:        0x10bd2db15 - __mh_execute_header
   5:        0x10bd2d8b7 - __mh_execute_header
   6:        0x10bd2dfcb - __mh_execute_header
   7:        0x10bd2abbc - __mh_execute_header
   8:        0x10bd2a979 - __mh_execute_header
   9:        0x10bd2dd42 - __mh_execute_header
  10:        0x10c556a45 - __mh_execute_header
  11:        0x10c556ace - __mh_execute_header
  12:        0x10b91b5c9 - __mh_execute_header
  13:        0x10bf9cf45 - __mh_execute_header
  14:        0x10bf71023 - __mh_execute_header
  15:        0x10bf9d724 - __mh_execute_header
  16:        0x10bf7093f - __mh_execute_header
  17:        0x10bf9d724 - __mh_execute_header
  18:        0x10bf71fc3 - __mh_execute_header
  19:        0x10bf9d7bb - __mh_execute_header
  20:        0x10bf71fc3 - __mh_execute_header
  21:        0x10bf99deb - __mh_execute_header
  22:        0x10bf65c82 - __mh_execute_header
  23:        0x10c1bfed7 - __mh_execute_header
  24:        0x10c1c3c9d - __mh_execute_header
  25:        0x10c09f680 - __mh_execute_header
  26:        0x10c0a033b - __mh_execute_header
  27:        0x10c0a097e - __mh_execute_header
  28:        0x10c0a0cc9 - __mh_execute_header
  29:        0x10c0a03b8 - __mh_execute_header
  30:        0x10c0a1a52 - __mh_execute_header
  31:        0x10c0a09bf - __mh_execute_header
  32:        0x10c0a097e - __mh_execute_header
  33:        0x10c0a0cc9 - __mh_execute_header
  34:        0x10c0a097e - __mh_execute_header
  35:        0x10c0a0cc9 - __mh_execute_header
  36:        0x10c0a097e - __mh_execute_header
  37:        0x10c0a0cc9 - __mh_execute_header
  38:        0x10c0a097e - __mh_execute_header
  39:        0x10c0a097e - __mh_execute_header
  40:        0x10c0a0cc9 - __mh_execute_header
  41:        0x10c0a097e - __mh_execute_header
  42:        0x10c0a0cc9 - __mh_execute_header
  43:        0x10c0a097e - __mh_execute_header
  44:        0x10c0a0cc9 - __mh_execute_header
  45:        0x10c0a097e - __mh_execute_header
  46:        0x10c0a0cc9 - __mh_execute_header
  47:        0x10c0a097e - __mh_execute_header
  48:        0x10c0a0cc9 - __mh_execute_header
  49:        0x10c0a097e - __mh_execute_header
  50:        0x10c0a0cc9 - __mh_execute_header
  51:        0x10c0a097e - __mh_execute_header
  52:        0x10c0a0cc9 - __mh_execute_header
  53:        0x10c0a097e - __mh_execute_header
  54:        0x10c0a0cc9 - __mh_execute_header
  55:        0x10c0a097e - __mh_execute_header
  56:        0x10c0a0cc9 - __mh_execute_header
  57:        0x10c0a097e - __mh_execute_header
  58:        0x10c0a0cc9 - __mh_execute_header
  59:        0x10c0a097e - __mh_execute_header
  60:        0x10c0a0cc9 - __mh_execute_header
  61:        0x10c0a097e - __mh_execute_header
  62:        0x10c0a0cc9 - __mh_execute_header
  63:        0x10c0a097e - __mh_execute_header
  64:        0x10c0a0cc9 - __mh_execute_header
  65:        0x10c0a097e - __mh_execute_header
  66:        0x10c0a0cc9 - __mh_execute_header
  67:        0x10c0a097e - __mh_execute_header
  68:        0x10c0a0cc9 - __mh_execute_header
  69:        0x10c0a097e - __mh_execute_header
  70:        0x10c20a51f - __mh_execute_header
  71:        0x10c179a78 - __mh_execute_header
  72:        0x10c25d14f - __mh_execute_header
  73:        0x10c32bbf5 - __mh_execute_header
  74:        0x10c3170ae - __mh_execute_header
  75:        0x10c2ca186 - __mh_execute_header
  76:        0x10c2cc82c - __mh_execute_header
  77:        0x10bd2dc64 - __mh_execute_header
  78:        0x10bd24a0e - __mh_execute_header
  79:        0x10c33066c - __mh_execute_header
  80:     0x7ff8043bf386 - <unknown>
```

I am not sure what is causing the error but I import cite and references from a BibTex file. One entry seems to be causing the issue:

```
@article{Redin2019Dec,
    author = "Redin, David and Frick, Tobias and Aghelpasand, Hooman and Käller, Max and Borgström, Erik and Olsen, Remi-Andre and Ahmadian, Afshin",
    title = "{High throughput barcoding method for genome-scale phasing}",
    journal = "Sci. Rep.",
    volume = "9",
    number = "18116",
    year = "2019",
    pages = "1--8",
    month = "December",
    issn = "2045-2322",
    publisher = "Nature Publishing Group",
    doi = "10.1038/s41598-019-54446-x"
}
``` 

Specifically the field `pages = "1--8",` as when I remove this field the file can magically compile without any errors. 

However if I try and create a small example (*.typ file that imports a *.bib) using the same BibTex entry it can compile without errors. 

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

Edit:
- Compiled a *.typ file not *.pdf.
- Clarify small example that did not reproduce the error

**Comment by pontushojer** [comment]
The same error appeared again in a new BibTex entry which also has a similar page field `pages = {1--4},`.

**Comment by Leedehai** [comment]
Interestingly, I was not able to repro this on macOS:

File 1: test.typ
```typ
A @Redin2019Dec.
B #cite(<Redin2019Dec>).

#bibliography("test.bib")
```

File 2: test.bib
```bib
@article{netwok,
  title={At-scale impact of the {Net Wok}: A culinarically holistic investigation of distributed dumplings},
  author={Astley, Rick and Morris, Linda},
  journal={Armenian Journal of Proceedings},
  volume={61},
  pages={192--219},
  year={2020},
  publisher={Automattic Inc.}
}

@article{Redin2019Dec,
    author = "Redin, David and Frick, Tobias and Aghelpasand, Hooman and Käller, Max and Borgström, Erik and Olsen, Remi-Andre and Ahmadian, Afshin",
    title = "{High throughput barcoding method for genome-scale phasing}",
    journal = "Sci. Rep.",
    volume = "9",
    number = "18116",
    year = "2019",
    pages = "1--8",
    month = "December",
    issn = "2045-2322",
    publisher = "Nature Publishing Group",
    doi = "10.1038/s41598-019-54446-x"
}
```

Either with the HEAD...
```sh
$ git rev-parse HEAD
7cb257a1accdf0af99b7210a2db53ea562764e15 # the most recent Typst checkout on the "main" branch

$ target/debug/typst compile test.typ # Succeeds; 'echo $?' prints 0
```
...OR with the released binary.
```sh
$ typst --version
typst 0.10.0 (70ca0d25)

$ typst compile test.typ # Succeeds; 'echo $?' prints 0
```

**Comment by Leedehai** [comment]
> However, if I try and create a small example using the same BibTex entry it can compile without errors.

Hi - a small typ file, or a small bib file?

**Comment by Enter-tainer** [comment]
```
typst compile thesis.pdf
```

I guess you want to compile a *.typ file?

**Comment by pontushojer** [comment]
> Hi - a small typ file, or a small bib file?

Both, I had the same setup. A main typ file that cites and imports a bib file.

> ```
> typst compile thesis.pdf
> ```
> 
> I guess you want to compile a *.typ file?

Yes it was a typ file. I modified the issue text.

**Comment by Enter-tainer** [comment]
https://github.com/typst/citationberg/blob/ed380070fa790379827ea2c2956e08eae7e5810d/src/lib.rs#L693C1-L694C1

related code

**Comment by Enivex** [comment]
> But i think 10^n % 2 can never be zero...

$10^n$ is divisible by $2$ for every $n \geq 1$

**Comment by Enter-tainer** [comment]
> > But i think 10^n % 2 can never be zero...
> 
> $10^n$ is divisible by $2$ for every $n \geq 1$, since $10 = 2 \cdot 5$ is.
> 

you are correct. my original post is very wrong. maybe it is a sufficiently large n that makes it overflow extactly to zero?

**Comment by Leedehai** [comment]
But if `pow()` overflows, I thought there should be an error `attempt to multiply with overflow`? I tested locally with Rust 1.75.0 as well as the [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a1d6a411794811773e4530972906c2f).

**Comment by Enter-tainer** [comment]
> But if `pow()` overflows, I thought there should be an error `attempt to multiply with overflow`? I tested locally with Rust 1.75.0 as well as the [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a1d6a411794811773e4530972906c2f).

you may try release mode

**Comment by DerDrodt** [comment]
Just a heads up that this issue is fixed in typst/citationberg#7. That PR (or a version of it) should be merged in the coming weeks.

**Comment by reknih** [comment]
Hey, I just did a minimal fix by performing the calculations in 64-bit integers and saturating the power operation. A larger overhaul of the page range mechanism is pending in https://github.com/typst/citationberg/pull/7.
